File: src/main/java/org/distril/beengine/util/SkinUtil.java
package org.distril.beengine.util;

import com.google.gson.JsonObject;
import com.nukkitx.protocol.bedrock.data.skin.*;

import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.Base64;
import java.util.List;

public class SkinUtil {

	public static SerializedSkin fromToken(JsonObject skinToken) {
		SerializedSkin.Builder builder = SerializedSkin.builder();

		if (skinToken.has("SkinId")) {
			builder.skinId(skinToken.get("SkinId").getAsString());
		}

		if (skinToken.has("PlayFabId")) {
			builder.playFabId(skinToken.get("PlayFabId").getAsString());
		}

		if (skinToken.has("SkinResourcePatch")) {
			builder.skinResourcePatch(new String(Base64.getDecoder().decode(skinToken.get("SkinResourcePatch").getAsString()),
					StandardCharsets.UTF_8));
		}

		builder.skinData(SkinUtil.getImageData(skinToken, "Skin"));

		if (skinToken.has("AnimatedImageData")) {
			List<AnimationData> animations = new ArrayList<>();
			var array = skinToken.getAsJsonArray("AnimatedImageData");
			array.forEach(element -> animations.add(getAnimationData(element.getAsJsonObject())));

			builder.animations(animations);
		}

		if (skinToken.has("PersonaSkin")) {
			builder.persona(skinToken.get("PersonaSkin").getAsBoolean());
			if (skinToken.has("PersonaPieces")) {
				List<PersonaPieceData> pieces = new ArrayList<>();
				var array = skinToken.getAsJsonArray("PersonaPieces");
				array.forEach(element -> pieces.add(SkinUtil.getPersonaPieceData(element.getAsJsonObject())));

				builder.personaPieces(pieces);
			}

			if (skinToken.has("PieceTintColors")) {
				List<PersonaPieceTintData> pieces = new ArrayList<>();
				var array = skinToken.getAsJsonArray("PieceTintColors");
				array.forEach(element -> pieces.add(SkinUtil.getPersonaPieceTintData(element.getAsJsonObject())));

				builder.tintColors(pieces);
			}
		}

		builder.capeData(SkinUtil.getImageData(skinToken, "Cape"));

		if (skinToken.has("SkinGeometryData")) {
			builder.geometryData(new String(Base64.getDecoder().decode(skinToken.get("SkinGeometryData").getAsString()),
					StandardCharsets.UTF_8));
		}

		if (skinToken.has("SkinAnimationData")) {
			builder.animationData(new String(Base64.getDecoder().decode(skinToken.get("SkinAnimationData").getAsString()),
					StandardCharsets.UTF_8));
		}

		if (skinToken.has("PremiumSkin")) {
			builder.premium(skinToken.get("PremiumSkin").getAsBoolean());
		}

		if (skinToken.has("CapeOnClassicSkin")) {
			builder.capeOnClassic(skinToken.get("CapeOnClassicSkin").getAsBoolean());
		}

		if (skinToken.has("CapeId")) {
			builder.capeId(skinToken.get("CapeId").getAsString());
		}

		if (skinToken.has("SkinColor")) {
			builder.skinColor(skinToken.get("SkinColor").getAsString());
		}

		if (skinToken.has("ArmSize")) {
			builder.armSize(skinToken.get("ArmSize").getAsString());
		}

		return builder.build();
	}

	private static AnimationData getAnimationData(JsonObject object) {
		var width = object.get("ImageWidth").getAsInt();
		var height = object.get("ImageHeight").getAsInt();
		var data = Base64.getDecoder().decode(object.get("Image").getAsString());
		var frames = object.get("Frames").getAsFloat();
		var textureType = AnimatedTextureType.values()[object.get("Type").getAsInt()];
		AnimationExpressionType expressionType = AnimationExpressionType.BLINKING;
		if (object.has("ExpressionType")) {
			expressionType = AnimationExpressionType.values()[object.get("ExpressionType").getAsInt()];
		}

		return new AnimationData(ImageData.of(width, height, data), textureType, frames, expressionType);
	}

	private static PersonaPieceData getPersonaPieceData(JsonObject skinToken) {
		var pieceId = skinToken.get("PieceId").getAsString();
		var pieceType = skinToken.get("PieceType").getAsString();
		var packId = skinToken.get("PackId").getAsString();
		var isDefault = skinToken.get("IsDefault").getAsBoolean();
		var productId = skinToken.get("ProductId").getAsString();

		return new PersonaPieceData(pieceId, pieceType, packId, isDefault, productId);
	}

	private static PersonaPieceTintData getPersonaPieceTintData(JsonObject skinToken) {
		var pieceType = skinToken.get("PieceType").getAsString();

		List<String> colors = new ArrayList<>();
		skinToken.getAsJsonArray("Colors").forEach(element -> colors.add(element.getAsString()));

		return new PersonaPieceTintData(pieceType, colors);
	}

	private static ImageData getImageData(JsonObject token, String name) {
		if (token.has(name + "Data")) {
			var skinImage = Base64.getDecoder().decode(token.get(name + "Data").getAsString());
			if (token.has(name + "ImageHeight") && token.has(name + "ImageWidth")) {
				var width = token.get(name + "ImageWidth").getAsInt();
				var height = token.get(name + "ImageHeight").getAsInt();
				return ImageData.of(width, height, skinImage);
			} else {
				return switch (skinImage.length / 4) {
					case 2048 -> ImageData.of(64, 32, skinImage);
					case 4096 -> ImageData.of(64, 64, skinImage);
					case 8192 -> ImageData.of(128, 64, skinImage);
					case 16384 -> ImageData.of(128, 128, skinImage);
					default -> ImageData.EMPTY;
				};
			}
		}

		return ImageData.EMPTY;
	}
}

File: src/main/kotlin/org/distril/beengine/util/SkinUtil.kt
package org.distril.beengine.util

import com.google.gson.JsonObject
import com.nukkitx.protocol.bedrock.data.skin.*
import java.nio.charset.StandardCharsets
import java.util.*

object SkinUtil {

	fun fromToken(skinJson: JsonObject): SerializedSkin {
		val builder = SerializedSkin.builder()

		builder.skinId(skinJson["SkinId"]?.asString)
		builder.playFabId(skinJson["PlayFabId"]?.asString)

		skinJson["SkinResourcePatch"]?.asString?.let {
			builder.skinResourcePatch(String(Base64.getDecoder().decode(it), StandardCharsets.UTF_8))
		}

		builder.skinData(this.getImageData(skinJson, "Skin"))

		val animations = skinJson.getAsJsonArray("AnimatedImageData")?.map {
			this.getAnimationData(it.asJsonObject)
		}
		builder.animations(animations)

		skinJson["PersonaSkin"]?.asBoolean?.let { personaSkin ->
			builder.persona(personaSkin)

			val pieces = skinJson.getAsJsonArray("PersonaPieces")?.map {
				this.getPersonaPieceData(it.asJsonObject)
			}
			builder.personaPieces(pieces)

			val tintColors = skinJson.getAsJsonArray("PieceTintColors")?.map {
				this.getPersonaPieceTintData(it.asJsonObject)
			}
			builder.tintColors(tintColors)
		}

		builder.capeData(this.getImageData(skinJson, "Cape"))

		skinJson["SkinGeometryData"]?.asString?.let {
			builder.geometryData(String(Base64.getDecoder().decode(it), StandardCharsets.UTF_8))
		}

		skinJson["SkinAnimationData"]?.asString?.let {
			builder.animationData(String(Base64.getDecoder().decode(it), StandardCharsets.UTF_8))
		}

		builder.premium(skinJson["PremiumSkin"]?.asBoolean == true)
		builder.capeOnClassic(skinJson["CapeOnClassicSkin"]?.asBoolean == true)
		builder.capeId(skinJson["CapeId"]?.asString)
		builder.skinColor(skinJson["SkinColor"]?.asString)
		builder.armSize(skinJson["ArmSize"]?.asString)

		return builder.build()
	}

	private fun getAnimationData(animationJson: JsonObject): AnimationData {
		val width = animationJson["ImageWidth"].asInt
		val height = animationJson["ImageHeight"].asInt
		val data = Base64.getDecoder().decode(animationJson["Image"].asString)
		val frames = animationJson["Frames"].asFloat
		val textureType = AnimatedTextureType.values()[animationJson["Type"].asInt]
		var expressionType = AnimationExpressionType.BLINKING
		animationJson["ExpressionType"]?.asInt?.let {
			expressionType = AnimationExpressionType.values()[it]
		}

		return AnimationData(ImageData.of(width, height, data), textureType, frames, expressionType)
	}

	private fun getPersonaPieceData(pieceJson: JsonObject): PersonaPieceData {
		val pieceId = pieceJson["PieceId"].asString
		val pieceType = pieceJson["PieceType"].asString
		val packId = pieceJson["PackId"].asString
		val isDefault = pieceJson["IsDefault"].asBoolean
		val productId = pieceJson["ProductId"].asString
		return PersonaPieceData(pieceId, pieceType, packId, isDefault, productId)
	}

	private fun getPersonaPieceTintData(pieceTintJson: JsonObject): PersonaPieceTintData {
		val pieceType = pieceTintJson["PieceType"].asString
		val colors = pieceTintJson.getAsJsonArray("Colors").map { it.asString }

		return PersonaPieceTintData(pieceType, colors)
	}

	private fun getImageData(imageJson: JsonObject, name: String): ImageData {
		val dataKey = name + "Data"
		val widthKey = name + "ImageWidth"
		val heightKey = name + "ImageHeight"

		if (imageJson.has(dataKey)) {
			val skinImage = Base64.getDecoder().decode(imageJson[dataKey].asString)
			val width = imageJson[widthKey]?.asInt
			val height = imageJson[heightKey]?.asInt

			return if (width != null && height != null) {
				ImageData.of(width, height, skinImage)
			} else {
				when (skinImage.size / 4) {
					2048 -> ImageData.of(64, 32, skinImage)
					4096 -> ImageData.of(64, 64, skinImage)
					8192 -> ImageData.of(128, 64, skinImage)
					16384 -> ImageData.of(128, 128, skinImage)
					else -> ImageData.EMPTY
				}
			}
		}

		return ImageData.EMPTY
	}
}
